### PR TITLE
Dynamic calculation of allocations

### LINKED ISSUE #20: Change service to use versioned target population * coverage * doses * wastage * plans instead of weekly levels
Requires https://github.com/fielded/nav-etl/issues/144 and https://github.com/fielded/nav-etl/issues/148

The `weeklyLevels` need to be calculated dynamically from 15th December 2016 but still use the db `weeklyLevels` until that date. The reason is that for 2016 `weeklyLevels` are not necessarily based on `targetPopulation` (they actually are for LGAs, but not for states or zones).

NB: the allocations for devices are calculated based on the allocations of relevant vaccines or devices. See https://github.com/fielded/nav-etl/blob/ba0b224b883a73837762016a0ec3a5ac1270a2d7/lib/integrators/allocations/calculate.js

### PR DESCRIPTION
**Pull request by patriciagarcia**
Closes #20

Comparing the current threshold values with the ones calculated dynamically:
- for 2016, the results are exactly the same
- for 2017, the Kano state satellite store device allocations will change slightly because the allocations imported from the Excel use different wastage coefficients (related: ticket to correct the Kano satellite store device allocations for 2016 https://github.com/fielded/nav-etl/issues/198, not a priority right now)

##### min/reOrder/max thresholds comparison for 2017 (following national dashboard product order)

```diff
< "NW","Kano","Kano State Store - Bichi",3520,7020,17520,6800,13600,34000,5150,10300,25750,2190,4370,10920,2190,4370,10920,2870,5730,14320,4070,8140,20350,5150,10300,25750,1530,3050,7630,23124,46247,115618,3502,7004,17510,450,899,2246,186,371,927,274,548,1370,3520,7020,17520,2190,4370,10920,2190,4370,10920
---
> "NW","Kano","Kano State Store - Bichi",3520,7020,17520,6800,13600,34000,5150,10300,25750,2190,4370,10920,2190,4370,10920,2870,5730,14320,4070,8140,20350,5150,10300,25750,1530,3050,7630,23124,46247,115618,3502,7004,17510,481,961,2402,193,386,964,287,574,1434,3520,7020,17520,2190,4370,10920,2190,4370,10920
315c315
< "NW","Kano","Kano State Store - Dawakin",4640,9280,23180,8480,16960,42380,5610,11210,28020,3200,6390,15970,3200,6390,15970,1850,3700,9240,4710,9420,23530,5604,11208,28016,2450,4900,12240,26593,53186,132963,4635,9270,23175,672,1344,3358,256,511,1278,324,647,1618,4640,9280,23180,3200,6390,15970,3200,6390,15970
---
> "NW","Kano","Kano State Store - Dawakin",4640,9280,23180,8480,16960,42380,5610,11210,28020,3200,6390,15970,3200,6390,15970,1850,3700,9240,4710,9420,23530,5604,11208,28016,2450,4900,12240,26593,53186,132963,4635,9270,23175,703,1405,3513,255,510,1275,338,676,1690,4640,9280,23180,3200,6390,15970,3200,6390,15970
324c324
< "NW","Kano","Kano State Store - Gwale",5660,11300,28240,16480,32960,82400,9270,18540,46350,3300,6600,16480,3300,6600,16480,3180,6350,15870,5570,11130,27810,9270,18540,46350,2770,5530,13810,36627,73254,183134,5645,11289,28223,676,1352,3379,297,594,1484,433,866,2163,5660,11300,28240,3300,6600,16480,3300,6600,16480
---
> "NW","Kano","Kano State Store - Gwale",5660,11300,28240,16480,32960,82400,9270,18540,46350,3300,6600,16490,3300,6600,16490,3180,6350,15870,5570,11130,27810,9270,18540,46350,2770,5530,13810,36627,73254,183134,5645,11289,28223,726,1451,3626,311,621,1553,455,910,2274,5660,11300,28240,3300,6600,16490,3300,6600,16490
338c338
< "NW","Kano","Kano State Store - Nassarawa",6180,12360,30900,12360,24720,61800,7210,14420,36050,3090,6180,15450,3090,6180,15450,3090,6180,15450,9890,19780,49440,7210,14420,36050,2480,4950,12360,36050,72100,180250,6180,12360,30900,649,1298,3245,340,680,1700,433,866,2163,6180,12360,30900,3090,6180,15450,3090,6180,15450
---
> "NW","Kano","Kano State Store - Nassarawa",6180,12360,30900,12360,24720,61800,7220,14430,36060,3090,6180,15450,3090,6180,15450,3090,6180,15450,9890,19780,49440,7212,14422,36052,2480,4950,12360,36050,72100,180250,6180,12360,30900,680,1360,3400,340,680,1700,455,909,2271,6180,12360,30900,3090,6180,15450,3090,6180,15450
340c340
< "NW","Kano","Kano State Store - Rano",5160,10320,25800,9280,18540,46360,6290,12580,31450,2760,5510,13780,2760,5510,13780,2230,4460,11150,6210,12410,31030,6290,12580,31446,2480,4960,12390,28997,57994,144983,5159,10317,25792,567,1133,2833,272,544,1360,351,701,1751,5160,10320,25800,2760,5510,13780,2760,5510,13780
---
> "NW","Kano","Kano State Store - Rano",5160,10320,25800,9280,18540,46360,6290,12580,31450,2760,5510,13780,2760,5510,13780,2230,4460,11150,6210,12410,31030,6290,12580,31446,2480,4960,12390,28997,57994,144983,5159,10317,25792,606,1212,3030,284,568,1419,368,736,1840,5160,10320,25800,2760,5510,13780,2760,5510,13780
353c353
< "NW","Kano","Kano State Store - Wudil",4160,8300,20720,11840,23680,59180,6020,12030,30070,3170,6330,15810,3170,6330,15810,1360,2720,6800,5700,11400,28480,6014,12028,30066,2210,4420,11040,27611,55221,138051,4141,8282,20703,649,1298,3245,219,437,1092,328,656,1638,4160,8300,20720,3170,6330,15810,3170,6330,15810
---
> "NW","Kano","Kano State Store - Wudil",4160,8300,20720,11840,23680,59180,6020,12030,30070,3170,6330,15810,3170,6330,15810,1360,2720,6800,5700,11400,28480,6014,12028,30066,2210,4420,11040,27610,55219,138046,4141,8282,20703,696,1391,3477,228,456,1139,344,687,1716,4160,8300,20720,3170,6330,15810,3170,6330,15810
```

### CODE REVIEW

**Comment by tlvince** [comment]
Could we run this against a subset of locations and diff against the current static values?

File: src/config/coefficients.json [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{
+  "versions": [
+    {
+      "version": 1,
+      "date": "2016-01-01",
```
Comment (tlvince) [inline src/config/coefficients.json:5]: Nit: should we change this to 2017 for clarity?

File: src/location-transformer.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/* global moment:false */
+import config from './config/config.json'
+import { find } from './utils.js'
+import calculateWeeklyLevels from './weekly-levels-calculator'
+
+// centralized for whenever we implement #16
+const somethingIsWrong = () => undefined
```
Comment (tlvince) [inline src/location-transformer.js]: Super nit: `() => {}` is a little more idiomatic noop

File: src/location-transformer.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  const allocations = getFactor(location.allocations, date)
+  return allocations && allocations.weeklyLevels
+}
+
+export default function transform (location, productCoefficients, date) {
+  const weeksOfStock = getWeeksOfStock(location, date)
+  if (!weeksOfStock) {
+    somethingIsWrong()
```
Comment (tlvince) [inline src/location-transformer.js]: `somethingIsWrong()` -> `return somethingIsWrong()`?

File: src/location-transformer.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  }
+  return factor
+}
+
+const getCoefficients = (productCoefficients, date) => {
+  if (!(productCoefficients && productCoefficients.versions && productCoefficients.versions.length)) {
+    somethingIsWrong()
+  }
+  const version = getFactor(productCoefficients.versions, date)
+  return version && version.coefficients
```
Comment (tlvince) [inline src/location-transformer.js]: This'll return fasly if version and/or `.coefficients` are undefined. Perhaps throw or call `somethingIsWrong`?

File: src/thresholds.service.js [hunk 5]
```diff
@@ -90,7 +13,9 @@ class ThresholdsService {
   // For zones the thresholds are based on the state store required allocation for
   // the week, that information is passed as an optional param (`requiredStateStoresAllocation`).
   // That param is only used for zones.
-  calculateThresholds (location, stockCount, products, requiredStateStoresAllocation = {}) {
+  //
+  // Passing the coefficientVersions as a param so that it can be adapted later to use the database doc
+  calculateThresholds (location, stockCount, products, requiredStateStoresAllocation = {}, productCoefficients = defaultCoefficients) {
```
Comment (tlvince) [inline src/thresholds.service.js:18]: 👍

File: src/thresholds.service.js [hunk 6]
```diff
@@ -103,44 +28,51 @@ class ThresholdsService {
       return
     }
 
-    const { weeklyLevels, weeksOfStock, targetPopulations } = getFactors(stockCount, location)
+    const transformedLocation = transform(location, productCoefficients, stockCount.date)
```
Comment (tlvince) [inline src/thresholds.service.js]: Nit: `transformedLocation` to me reads like it'll be a copy of the full location, with some modified properties, whereas, `transform` returns a subset of props. Maybe `transformedLocation` -> `factors` or `locationFactors`?

File: src/thresholds.service.js [hunk 7]
```diff
@@ -103,44 +28,51 @@ class ThresholdsService {
...
+        presentation = parseInt(product.presentation, 10)
+      }
 
+      index[productId] = Object.keys(weeksOfStock).reduce((productThresholds, threshold) => {
+        const level = weeklyLevel * weeksOfStock[threshold]
         const roundedLevel = Math.ceil(level / presentation) * presentation
         productThresholds[threshold] = roundedLevel
 
-        if (location.level === 'zone' && requiredStateStoresAllocation[productId]) {
+        if (transformedLocation.level === 'zone' && requiredStateStoresAllocation[productId]) {
```
Comment (tlvince) [inline src/thresholds.service.js]: Why does `transform` duplicate the `level`? Can't the one on `location` be used?

**Review by tlvince — APPROVED** [review]
Code lgtm (just style/nits). Feel free to merge if you've already checked the results match up with the static values already in the docs/or what you're expecting them to be (after merging this in, the static versions will just be ignored, right?).

File: src/config/coefficients.json (hunk 1 above)
Comment (patriciagarcia) [reply src/config/coefficients.json:5]: mm, I'd leave it like that, it gives more information about where this value came from (I am guessing it came from an Excel dashboard with that date?)

File: src/location-transformer.js (hunk 2 above)
Comment (patriciagarcia) [reply src/location-transformer.js]: ah, good one!

File: src/location-transformer.js (hunk 3 above)
Comment (patriciagarcia) [reply src/location-transformer.js]: oops, that's a very good one, otherwise it won't break the execution..

File: src/thresholds.service.js (hunk 6 above)
Comment (patriciagarcia) [reply src/thresholds.service.js]: My original idea was to have use the location transformer as a normalization layer that can understand and unify different location schemas (therefore also the duplication of `level` mentioned below), but, the way it turned out at the end, I think your idea of just having it as a factor extractor is better.

**Comment by patriciagarcia** [comment]
@tlvince, I implement that logging to the console we talked about, can you have a look to see if it looks good? (added a follow up issue for Google Analytics logging https://github.com/fielded/angular-nav-thresholds/issues/28)

**Comment by tlvince** [comment]
LGTM.

**Pull request merged**